packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -341,20 +336,31 @@ function shouldUpgradeListener(
 
 export function listenToNativeEvent(
   topLevelType: DOMTopLevelEventType,
-  target: EventTarget,
-  listenerMap: ElementListenerMap,
-  eventSystemFlags: EventSystemFlags,
   isCapturePhaseListener: boolean,
+  rootContainerElement: EventTarget,
+  targetElement?: Element,
   isPassiveListener?: boolean,
-  priority?: EventPriority,
+  listenerPriority?: EventPriority,
+  eventSystemFlags?: EventSystemFlags = PLUGIN_EVENT_SYSTEM,
 ): void {
+  let target = rootContainerElement;
   // TOP_SELECTION_CHANGE needs to be attached to the document
   // otherwise it won't capture incoming events that are only
   // triggered on the document directly.
   if (topLevelType === TOP_SELECTION_CHANGE) {
-    target = (target: any).ownerDocument || target;
-    listenerMap = getEventListenerMap(target);
+    target = (rootContainerElement: any).ownerDocument;
+  }
+  // If the event can be delegated, we can register it to the root container.
+  // Otherwise, we should register the event to the target element.
+  if (targetElement !== undefined && nonDelegatedEvents.has(topLevelType)) {
+    eventSystemFlags |= IS_NON_DELEGATED;
+    target = targetElement;
+  }
+  // This is temporary, we will remove this when introducing focusin/focusout
```

Comment:
Let's stack this PR on https://github.com/facebook/react/pull/19186 and remove special cases.